### ISSUE TITLE
Cannot cast `float literal` to sum type

### ISSUE THREAD

**Issue by ulises-jeremias** [vlang/v#7851]
**V version:** V 0.2.1 90839a1.43adbf4
**OS:** linux, "Arch Linux"

**What did you do?**

I used to have this type, when `any_int` and `any_float` were not private:
```v
// `Num` is a sum type that lists the possible types to be used to define tensor values
pub type Num = any_int | any_float | byte | f32 | f64 | i16 | i64 | i8 | int | u16 | u32 |
	u64
```

and a function as follow

```v
fn foo(x Num) {
. . .
}
```

and I was able to use it as `foo(4.0)`.

Then, I redefined the type `Num` as:
```v
// `Num` is a sum type that lists the possible types to be used to define tensor values
pub type Num = byte | f32 | f64 | i16 | i64 | i8 | int | u16 | u32 |
	u64
```

Now, when executing `foo(4.0)` it fails.

**What did you expect to see?**

It works.
 
**What did you see instead?**

error: cannot use `float literal` as `Num` in argument 1 to `foo`. It is blocking the development of https://github.com/vlang/vtl

**Comment by UweKrueger** [comment]
The point is that `0.4` could either be `f64` or `f32`. A variable of a sum type always has to have one "real" type and the compiler refuses to assign one since it would be ambiguous - this is not a bug.
However I will add code to do automatic conversions to `int`/`f64` in this case as is already done during variable initialization.

This is essentially a copy of #6575 so I close this one here.

**Comment by UweKrueger** [comment]
This should now be fixed in bf904c2 by #7864.
